docs/rfc/rfc-7.adoc

```diff
@@ -0,0 +1,205 @@
+signing protocol.
+
+==== Timeouts
+
+The sweep should be completed in 4 hours to leave the floor for redemptions.
+\This time includes retries and a potential increase of the Bitcoin transaction
+fee. The signing protocol takes 72 blocks. Assuming an average block time of 12
+seconds, it takes approximately 15 minutes to execute the protocol. With 3
+malicious members in a signing group, we need 10 retries of the protocol in the
+worst case, so 2h 30 min (`P = (97 choose 51) / (100 choose 51) = 0.11393939393`).
+Thus, the timeout for signing a sweep transaction should be set to 2h 30 minutes
+leaving 60 minutes for a potential fee increase.
+
+Note that if the active wallet is permanently struggling with signing and can't
+produce a signature after 10 retries, it is better to not accept new deposits to
+that wallet and let another active wallet be created.
+
+==== Retries
+
+The signing protocol should be extended by an additional phase at the very
+beginning of each retry that will inform about the data being signed. For
+sweeping, it will inform about the Bitcoin sweep transaction proposed by the
+Coordinator. All signing members should broadcast this information if they have
+already validated the proposition. The signing group members continue with the
+protocol only if there are at least 51 messages informing them about the
+transaction and its successful validation. Group members who have turned on
+their nodes can learn about the transaction the rest of the group wants to sign
+and they can join the signing protocol if necessary.
+
+==== Increasing fee
+
+At any point in time, the current Coordinator (who is not necessarily the same
+Coordinator who proposed the batch) can propose increasing the Bitcoin fee for
+the sweep transaction that is in the mempool. The signing group members sign the
+transaction if the original transaction is in the mempool for at least 30
+minutes and the fee increase does not exceed the maximum one allowed by the
+Bridge. Given that the wallet is blocked with any other action until the sweep
+transaction is in the mempool, the fee bump signing protocol retries until the
+signature is produced.
+
+== Future Work
+
+The link:https://github.com/keep-network/tbtc-v2/pull/374[sparse relay] provides
+true proof of inclusion but requires every 6th block in the Bitcoin network to
+be recorded on Ethereum. We started with a proof of work in tBTC v2 Bridge and
+the light relay but one day we may switch to the sparse relay. This would allow
+establishing a reference between a point in time on Ethereum and Bitcoin chains.
+The reference point in time on Ethereum is a fixed block number - for example,
+we sweep every 100th block - and the reference point in time on Bitcoin is the
+last Bitcoin block recorded on Ethereum's sparse relay at the given Ethereum's
+block number.
+
+== Open Questions
+
+Is the maximum of 500 transactions in a batch enough to not risk not being able
```

Comment:
One important thing just came to my mind. We must remember that when we talk about signing a sweep transaction, we are actually talking about signing the N inputs (i.e. unlocking the deposits' UTXOs) of the sweep transaction. That means we need N independent instances of the signing protocol to be done. This is a major performance challenge we must be aware of. It also opens some complex scenarios like, what should we do if only part of N signatures were produced?

Reply:
That's a good point, I reworked the approach in https://github.com/keep-network/tbtc-v2/pull/414/commits/b057fefa249900c5e31b7166316bca3cd0839d50.

Every input of the Bitcoin sweep transaction needs to be signed separately. Given that each message being signed contains information about all other inputs and the output, the wallet must produce all signatures. Instead of giving up
after 2 hours, the wallet should block all other operations and try for at least a week. All signatures are done simultaneously, in separate instances of the signing state machine. This requires limiting the number of inputs being swept to 20 initially.